### PR TITLE
[SPARK-38474][CORE] Use error class in org.apache.spark.security

### PR DESCRIPTION
**Pull request by sandeep-katta**
### Why are the changes needed?

Refactored exceptions thrown in the package `org.apache.spark.security` in to be in inline Spark error class framework

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Modified impacted testcases and added new test cases

### CODE REVIEW

**Comment by sandeep-katta** [comment]
cc @bozhang2820

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 1]
```diff
@@ -325,4 +325,13 @@ object SparkCoreErrors {
...
   }
+
+  def authenticationFailedError(hostType: String): Throwable = {
+    new SparkException(errorClass = "AUTHENTICATION_FAILED",
+      messageParameters = Array(hostType), cause = null)
+  }
+
+  def streamClosed(): Throwable = {
+    new SparkException("Cipher stream is closed.")
+  }
```
Comment (bozhang2820) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:336]: Let's create one for this in error-classes.json as well.

File: core/src/main/scala/org/apache/spark/security/SocketAuthHelper.scala [hunk 2]
```diff
@@ -60,7 +61,7 @@ private[spark] class SocketAuthHelper(val conf: SparkConf) {
           shouldClose = false
         } else {
           writeUtf8("err", s)
-          throw new IllegalArgumentException("Authentication failed.")
+          throw SparkCoreErrors.authenticationFailedError("client")
```
Comment (bozhang2820) [inline core/src/main/scala/org/apache/spark/security/SocketAuthHelper.scala:64]: Not sure if we should change the exception type here. CC @MaxGekk for more guidance.

Comment (bozhang2820) [reply core/src/main/scala/org/apache/spark/security/SocketAuthHelper.scala:64]: @cloud-fan, are we allowed to change the exception type here?

Comment (bozhang2820) [reply core/src/main/scala/org/apache/spark/security/SocketAuthHelper.scala:64]: Also let's not forget to update the `@throws` in the scaladoc, in both `authClient` and `authToServer`.

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 3]
```diff
@@ -325,4 +325,14 @@ object SparkCoreErrors {
...
       messageParameters = Array(missingProperty), cause = null)
   }
+
+  def authenticationFailedError(hostType: String): Throwable = {
+    new SparkIllegalArgumentException(errorClass = "AUTHENTICATION_FAILED",
+      messageParameters = Array(hostType))
+  }
+
+  def streamClosed(streamType: String): Throwable = {
+    new SparkException(errorClass = "STREAM_CLOSED",
```
Comment (bozhang2820) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:335]: Use `SparkIOException` here as well?

**Comment by bozhang2820** [comment]
Change LGTM. Not sure about the GitHub Action test failures (about doc generations and docker integration tests). @HyukjinKwon any idea?

**Comment by HyukjinKwon** [comment]
yeah that test shouldn't be related, and it's just flaky.

File: core/src/test/scala/org/apache/spark/security/SocketAuthHelperSuite.scala [hunk 4]
```diff
@@ -16,13 +16,17 @@
...
 
-import org.apache.spark.{SparkConf, SparkFunSuite}
+import org.mockito.ArgumentMatchers.any
+import org.mockito.Mockito.{mock, when}
+
+import org.apache.spark.{SparkConf, SparkFunSuite, SparkIllegalArgumentException}
 import org.apache.spark.internal.config._
 import org.apache.spark.util.Utils
 
+
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/security/SocketAuthHelperSuite.scala:29]: Let's remove unrelated changes.

File: core/src/test/scala/org/apache/spark/security/SocketAuthHelperSuite.scala [hunk 5]
```diff
@@ -55,6 +62,23 @@ class SocketAuthHelperSuite extends SparkFunSuite {
     }
   }
 
+  test("failed to authenticate client") {
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/security/SocketAuthHelperSuite.scala:65]:
```suggestion
  test("SPARK-38474: failed to authenticate client") {
```

File: core/src/test/scala/org/apache/spark/security/CryptoStreamUtilsSuite.scala [hunk 6]
```diff
@@ -185,10 +185,12 @@ class CryptoStreamUtilsSuite extends SparkFunSuite {
...
     }
 
-    val e = intercept[IOException] {
+    val e = intercept[SparkIOException] {
       errorHandler.read(out)
     }
-    assert(e.getMessage().contains("is closed"))
+    assert(e.getErrorClass === "STREAM_CLOSED")
+    assert(e.getMessage ===
+            "[STREAM_CLOSED] Cipher stream is closed.")
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/security/CryptoStreamUtilsSuite.scala:193]:
```suggestion
    assert(e.getMessage === "[STREAM_CLOSED] Cipher stream is closed.")
```

**Comment by HyukjinKwon** [comment]
LGTM2 otherwise. @sandeep-katta you might need to rebase your branch and sync to the latest master branch to avoid the linter job failure.

File: core/src/main/scala/org/apache/spark/security/SocketAuthHelper.scala (hunk 2 above)
Comment (bozhang2820) [reply core/src/main/scala/org/apache/spark/security/SocketAuthHelper.scala:64]: @sandeep-katta ^^

Comment (sandeep-katta) [reply core/src/main/scala/org/apache/spark/security/SocketAuthHelper.scala:64]: It is still `IllegalArgumentException`, don't I need to update scaladoc with `SparkIllegalArgumentException` ?

Comment (bozhang2820) [reply core/src/main/scala/org/apache/spark/security/SocketAuthHelper.scala:64]: I see your point. I'm okay with keeping the current ones.

**Review by bozhang2820 — APPROVED** [review]

**Comment by bozhang2820** [comment]
@cloud-fan @MaxGekk could you review this?

File: core/src/main/resources/error/error-classes.json [hunk 7]
```diff
@@ -341,6 +346,11 @@
     ],
     "sqlState" : "22023"
   },
+  "STREAM_CLOSED" : {
+    "message" : [
+      "<streamType> stream is closed."
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:351]: This doesn't really like a user-facing error. I think IO errors are all internal errors? cc @srielau

Comment (sandeep-katta) [reply core/src/main/resources/error/error-classes.json:351]: @srielau could you confirm please, based on the update I will update the changes

Comment (srielau) [reply core/src/main/resources/error/error-classes.json:351]: I can't make that call. Why could the stream be closed? Would you expect a Jira ticket to be opened to investigate this error? What is the customer action? What is the Databricks action....
